internal/monitors/elasticsearch/client/client.go

```diff
@@ -0,0 +1,181 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+const (
+	nodeStatsEndpoint          = "_nodes/_local/stats/transport,http,process,jvm,indices,thread_pool"
+	clusterHealthStatsEndpoint = "_cluster/health"
+	nodeInfoEndpoint           = "_nodes/_local"
+	masterNodeEndpoint         = "_cluster/state/master_node"
+	allIndexStatsEndpoint      = "_all/_stats"
+)
+
+type esClient struct {
+	host       string
+	port       string
+	scheme     string
+	username   string
+	password   string
+	httpClient *http.Client
+}
+
+// ESHttpClient holds methods hitting various ES stats endpoints
+type ESHttpClient interface {
+	GetNodeAndThreadPoolStats() (*NodeStatsOutput, error)
+	GetClusterStats() (*ClusterStatsOutput, error)
+	GetNodeInfo() (*NodeInfoOutput, error)
+	GetMasterNodeInfo() (*MasterInfoOutput, error)
+	GetIndexStats() (*IndexStatsOutput, error)
+}
+
+// NewESClient creates a new esClient
+func NewESClient(host string, port string, useHTTPS bool, username string, password string) ESHttpClient {
+	httpClient := &http.Client{}
+	scheme := "http"
+
+	if useHTTPS {
+		scheme = "https"
+	}
+
+	return &esClient{
+		host:       host,
+		port:       port,
+		scheme:     scheme,
+		username:   username,
+		password:   password,
+		httpClient: httpClient,
+	}
+}
+
+// Method to collect index stats
+func (c *esClient) GetIndexStats() (*IndexStatsOutput, error) {
+	url := fmt.Sprintf("%s://%s:%s/%s", c.scheme, c.host, c.port, allIndexStatsEndpoint)
```

Comment:
This is a lot of repeated code.  You can just have a method that decodes the json to a target object like this:

```
func (c *esClient) fetchJSON(url string, obj interface{}) error {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("could not get url %s: %v", url, err)
	}

	req.SetBasicAuth(username, password)
	res, err := httpClient.Do(req)

	if err != nil {
		return nil, fmt.Errorf("could not get url %s: %v", url, err)
	}
	defer res.Body.Close()

	if res.StatusCode != 200 {
		return nil, fmt.Errorf("Received status code that's not 200: %s, url: %s", string(body), url)
	}

       err := json.NewDecoder(res.Body).Decode(obj)
	if err != nil {
		return nil, fmt.Errorf("could not get url %s: %v", url, err)
	}

	return nil
}
```

And then call it like this:

```
func (c *esClient) GetIndexStats() (*IndexStatsOutput, error) {
     	url := fmt.Sprintf("%s://%s:%s/%s", c.scheme, c.host, c.port, allIndexStatsEndpoint)
        var indexStatsOutput IndexStatsOutput
        return &indexStatsOutput, c.fetchJSON(url, &indexStatsOutput)
}
```

That would eliminate most of the boilerplate and also avoid having to copy the response body to a temporary variable as well, which would help reduce GC churn possibly.

Reply:
That's a good point. I've used this pattern now.